### PR TITLE
[DBT-43] Parsed Csv By Creating Csv Method

### PR DESCRIPTION
**Pull request by Ammar022**
- Created a Post request which receives a local file path in json body.
- Read and parse CSV.
- Return Data in JSON Format.
- Parsed Csv by creating Csv Method.

### CODE REVIEW

File: csv/csv.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"os"
+	"strings"
+)
+
+func Parsecsv(file *os.File) {
+	// define a list of file paths flag
+	filePaths := flag.String("files", "", "Comma-separated list of file paths")
+
+	// parse the command-line arguments
+	flag.Parse()
```
Comment (mzeeshanali) [inline csv/csv.go:17]: do we need this now, since we are getting path from request body?

File: main.go [hunk 2]
```diff
@@ -105,3 +72,40 @@ func helloWithName(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	w.Write(jsonRes)
 }
+
+func csvParsing(w http.ResponseWriter, r *http.Request) {
```
Comment (mzeeshanali) [inline main.go:76]: Please read the instructions again. This parse should be in the new package `csv`.

File: main.go [hunk 3]
```diff
@@ -20,44 +19,11 @@ type Response struct {
...
-		s, err := json.Marshal(records)
-		if err != nil {
-			fmt.Println("Failed to marshal data in json:", err)
-			return
-		}
-		fmt.Println(string(s))
-	}
+type FilePath struct {
+	Filepath string `json:"filepath"`
+}
```
Comment (mzeeshanali) [inline main.go:24]: this shouldn't be in main

**Review by mzeeshanali — CHANGES_REQUESTED** [review]
Please update your README.md. It's not properly formatted.

File: main.go (hunk 2 above)
Comment (Ammar022) [reply main.go:76]: okay got it. onto it.

File: csv/csv.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"encoding/csv"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+)
+
+type FilePath struct {
```
Comment (mzeeshanali) [inline csv/csv.go:11]:
```suggestion
type filePath struct {
```

File: csv/csv.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	Filepath string `json:"filepath"`
+}
+
+func ParseCsv(w http.ResponseWriter, r *http.Request) {
```
Comment (mzeeshanali) [inline csv/csv.go:15]: add comment about what this function is about

File: csv/csv.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	var filePath FilePath
```
Comment (mzeeshanali) [inline csv/csv.go:22]:
```suggestion
	var filePath filePath
```

File: csv/csv.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+		return
+	}
+
+	file, err := os.Open(filePath.Filepath)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	defer file.Close()
+	reader := csv.NewReader(file)
```
Comment (mzeeshanali) [inline csv/csv.go:35]: add line here

File: csv/csv.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+	res, err := json.Marshal(records)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(res)
+
```
Comment (mzeeshanali) [inline csv/csv.go:50]: remove this empty line

File: README.md [hunk 9]
```diff
@@ -4,10 +4,10 @@
 - Http get methods that say hello in "JSON Format"
 
 ##How to execute
```
Comment (mzeeshanali) [inline README.md:6]:
```suggestion
## How to execute
```

File: README.md [hunk 10]
```diff
@@ -4,10 +4,10 @@
...
-- Open terminal and execute: go run main.go
-- Open terminal and execute: go run main.go -files=filename.csv if you want to parse Csv
-- Open: http://localhost:8080/hello/{name} to get result with name
-- Open: http://localhost:8080/hello to get resulut without name
+- Open terminal and execute: go run main.go.
+- http://localhost:8080/hello/{name} to get result with name.
+- http://localhost:8080/hello to get resulut without name.
+- http://localhost:8080/csv with {"filepath": "/path/to/file"} for post request.
 
 ##Test case
```
Comment (mzeeshanali) [inline README.md:12]:
```suggestion
## Test case
```

**Review by mzeeshanali — CHANGES_REQUESTED** [review]
Please add comments in your code.

File: csv/csv.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	"io/ioutil"
+	"net/http"
+	"os"
+)
+
+type filePath struct {
+	Filepath string `json:"filepath"`
+}
+
+// it will receive path to csv file as request body and in return print data in JSON Format
```
Comment (mzeeshanali) [inline csv/csv.go:15]:
```suggestion
// ParseCsv receives path to csv file in request body and prints data in JSON format
```

File: csv/csv.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	res, err := json.Marshal(records)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(res)
```
Comment (mzeeshanali) [inline csv/csv.go:52]: comments are still missing in the code.

**Review by mzeeshanali — CHANGES_REQUESTED** [review]

File: csv/csv.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	err = json.Unmarshal(body, &filePath)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// get & store file from path given in request body
+	file, err := os.Open(filePath.Filepath)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (mzeeshanali) [inline csv/csv.go:35]:
```suggestion
		http.Error(w, err.Error(), http.InternalServerErro)
```

File: csv/csv.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+		return
+	}
+	defer file.Close()
+
+	// reading file
+	reader := csv.NewReader(file)
+
+	records, err := reader.ReadAll()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (mzeeshanali) [inline csv/csv.go:45]:
```suggestion
		http.Error(w, err.Error(), http.InternalServerError)
```

File: csv/csv.go [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+		return
+	}
+
+	// converting file data in JSON format to write as output
+	res, err := json.Marshal(records)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
```
Comment (mzeeshanali) [inline csv/csv.go:52]:
```suggestion
		http.Error(w, err.Error(), http.InternalServerError)
```

**Review by mzeeshanali — CHANGES_REQUESTED** [review]

File: csv/csv.go [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// converting file data in JSON format to write as output
+	res, err := json.Marshal(records)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```
Comment (mzeeshanali) [inline csv/csv.go:54]: also log errors.

**Review by mzeeshanali — CHANGES_REQUESTED** [review]

File: main.go [hunk 17]
```diff
@@ -92,13 +59,14 @@ func helloWithName(w http.ResponseWriter, r *http.Request) {
...
-		Message:   "Welcome" + name + "!",
+		Message:   "Welcome " + name + "!",
 		Timestamp: time.Now().Format(time.RFC3339),
 	}
 
 	// converting data to json format
 	jsonRes, err := json.Marshal(res)
 	if err != nil {
+		log.Println(err)
 		fmt.Println("Failed to marshal data in json:", err)
```
Comment (mzeeshanali) [inline main.go:70]:
```suggestion
		log.Println("Failed to marshal data in json:", err)
```

File: main.go [hunk 18]
```diff
@@ -80,6 +46,7 @@ func helloWithoutName(w http.ResponseWriter, r *http.Request) {
...
```
Comment (mzeeshanali) [inline main.go:50]:
```suggestion
		log.Println("Failed to marshal data in json:", err)
```

**Review by mzeeshanali — CHANGES_REQUESTED** [review]

File: csv/csv.go [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+// ParseCsv receives path to csv file in request body and prints data in JSON format
+func ParseCsv(w http.ResponseWriter, r *http.Request) {
+	// read request body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Println(err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
```
Comment (mzeeshanali) [inline csv/csv.go:24]:
```
log.Printf("Request body:\t%s\n", string(body))
```

please log request body too.

**Review by mzeeshanali — CHANGES_REQUESTED** [review]

**Review by mzeeshanali — APPROVED** [review]

**Pull request merged**